packages/grafana-data/src/transformations/transformers/insertValues.ts

```diff
@@ -0,0 +1,141 @@
+    value: null,
+    isRefField: IS_TIME_FIELD,
+    guessInterval: false,
+  },
+  operator: ({ mode, value = null, isRefField = IS_TIME_FIELD, guessInterval = false, threshold = THRESHOLD_HOUR }) => (
+    source
+  ) =>
+    source.pipe(
+      map((dataFrames) => {
+        if (mode === InsertValuesMode.Interval) {
+          return dataFrames.map((f) => insertValuesInterval(f, value ?? null, guessInterval ?? false, isRefField));
+        } else {
+          return dataFrames.map((f) => insertValuesThreshold(f, value ?? null, threshold, isRefField ?? IS_TIME_FIELD));
+        }
+      })
+    ),
+};
+
+// expects ascending number values in refField
+function getInterval(refField: Field, guessFromValues = false) {
+  let interval = Math.max(refField.config?.interval ?? 0, 0);
+
+  if (interval === 0 && guessFromValues && refField.values.length > 2) {
+    let refVals = refField.values.toArray();
+
+    interval = Infinity;
+
+    for (let i = 1; i < refVals.length; i++) {
+      interval = Math.min(interval, refVals[i] - refVals[i - 1]);
+    }
+  }
+
+  return Number.isFinite(interval) ? interval : 0;
+}
+
+// assumes some interval steps may be absent, but what remains in the time field:
+// 1. contains only numbers, would pass Number.isFinite(value)
+// 2. conforms to the specified interval
+// 3. is in ascending order
+export function insertValuesInterval(
+  frame: DataFrame,
+  value: any = null,
+  guessFromValues = false,
+  isRefField = IS_TIME_FIELD
+) {
+  if (frame.length <= 1) {
+    return frame;
+  }
+
+  let refField = frame.fields.find(isRefField);
+
+  if (refField == null) {
+    return frame;
+  }
+
+  let interval = getInterval(refField, guessFromValues);
+
+  if (interval === 0) {
+    return frame;
+  }
+
+  let refValues = refField.values.toArray();
+  let len = refValues.length;
+  let minValue = refValues[0];
+  let maxValue = refValues[len - 1];
+  let filledLen = (maxValue - minValue) / interval + 1;
+
+  // cheap exit length matches expected length
+  if (len === filledLen) {
+    return frame;
+  }
+
+  let filledFieldValues: any[][] = [];
+
+  for (let field of frame.fields) {
+    let filledValues = Array(filledLen);
+
+    if (field === refField) {
+      for (let i = 0; i < filledLen; i++) {
+        filledValues[i] = minValue + i * interval;
+      }
+    } else {
+      let fieldValues = field.values.toArray();
+
+      for (let i = 0, j = 0; i < filledLen; i++) {
+        filledValues[i] = refValues[j] === minValue + i * interval ? fieldValues[j++] : value;
+      }
+    }
+
+    filledFieldValues.push(filledValues);
+  }
+
+  let outFrame: DataFrame = {
+    ...frame,
+    length: filledLen,
+    fields: frame.fields.map((field, i) => ({
+      ...field,
+      values: new ArrayVector(filledFieldValues[i]),
+    })),
+  };
+
+  return outFrame;
+}
+
+// TODO: implement
+function insertValuesThreshold(frame: DataFrame, value: any = null, threshold: number, isRefField = IS_TIME_FIELD) {
```

Comment:
what would be the difference in behavior for this one?

Reply:
this one is the opposite of "Connect null values < threshold", it will insert nulls (disconnect/insert gaps) between adjacent points if the refField delta is > threshold. it does not take interval into account at all.

let's say you have this originally

```
time:  [  1,  2,  3, 10, 11, 12, 15]
value: [  1,  1,  1,  1,  1,  1,  1]
```

with a threshold of 5, you'll end up with

```
time:  [  1,  2,  3,  6,    10, 11, 12, 15]
value: [  1,  1,  1,  null,  1,  1,  1,  1]
```

where the `null` gap is inserted roughly between 3 and 10, cause 10 - 3 >= 5

Reply:
@leeoniya so the main difference is that `insertValuesInterval` could insert multiple `null` values if the gap spans multiple intervals, and `insertValuesThreshold`  will insert at most 1 `null`? Is there a benefit to having both?

Reply:
sorry, should have been more clear. the `threshold` mode would do the the more expensive operation, which is to detect gaps in any irregular data, rather than assuming the data already follows the specified interval. so `interval` mode can be viewed as an optimized (and much more common) subset of `threshold` mode.

i'll see how fast i can get `threshold` working today and if it's not too much slower then we can always use it. i have my doubts, though!

Reply:
updated code to always work in `threshold` mode, where threshold is taken from the time field's `config.interval`, when present.